File: problems/605/605.can-place-flowers.py
#
# @lc app=leetcode id=605 lang=python3
#
# [605] Can Place Flowers
#


# @lc code=start
class Solution:
    def canPlaceFlowers(self, flowerbed: List[int], n: int) -> bool:
        if n == 0:
            return True

        for i in range(len(flowerbed)):
            if flowerbed[i]:
                continue

            valid = True
            if i > 0 and flowerbed[i - 1]:
                valid = False
            if i < len(flowerbed) - 1 and flowerbed[i + 1]:
                valid = False

            if valid:
                flowerbed[i] = 1
                n -= 1

            if n <= 0:
                return True

        return False


# @lc code=end

File: problems/605/605.can-place-flowers.go
/*
 * @lc app=leetcode id=605 lang=golang
 *
 * [605] Can Place Flowers
 */

package solution

// @lc code=start
import (
	"fmt"
)

func canPlaceFlowers(flowerbed []int, n int) bool {
	if n == 0 {
		return true
	}

	i := 0
	for i < len(flowerbed) {
		if flowerbed[i] == 1 {
			i += 2
			continue
		}

		valid := i == 0 || flowerbed[i-1] == 0
		valid = valid && (i == len(flowerbed)-1 || flowerbed[i+1] == 0)
		fmt.Println(i, valid)

		if valid {
			i += 2
			n -= 1
		} else {
			i += 1
		}

		if n <= 0 {
			return true
		}
	}

	return false
}

// @lc code=end
